### PR TITLE
[WIP] Implement exercise bowling

### PR DESCRIPTION
**Pull request by aes421**
https://github.com/exercism/problem-specifications/tree/master/exercises/bowling

### CODE REVIEW

**Comment by m-a-ge** [comment]
@aes421 build failed due to some code style issues:
```text
2.65s$ flake8
./exercises/bowling/bowling_test.py:73:1: E305 expected 2 blank lines after class or function definition, found 1
./exercises/bowling/example.py:13:9: E265 block comment should start with '# '
./exercises/bowling/example.py:25:17: E265 block comment should start with '# '
```

**Comment by m-a-ge** [comment]
@aes421 you forgot to add the exercise to `config.json`:
```text
-> An implementation for 'bowling' was found, but config.json does not reference this exercise.
```

**Comment by aes421** [comment]
Hey @m-a-ge I'm actually not done writing the test/implementation of this exercise yet.  I left the config stuff until last for me previous PR, but I can go ahead and do it now if you'd like.

**Comment by m-a-ge** [comment]
Ah, sorry, my bad. Ping me when you're done.

Side note: I don't usually leave comments in `WIP` PRs, but some people forget to remove the tag and it's unclear whether PR is done or not

**Comment by aes421** [comment]
@m-a-ge  done with this one.  Let me know what you'd like to see changed.

File: exercises/bowling/.gitignore [hunk 1]
```diff
@@ -0,0 +1 @@
+.cache/
```
Comment (m-a-ge) [inline exercises/bowling/.gitignore]: I would advise not to include this in repo `.gitignore`. You can your local one for this purpose

File: exercises/bowling/bowling_test.py [hunk 2]
```diff
@@ -0,0 +1,261 @@
+import unittest
+
+from example import BowlingGame
+
+
```
Comment (m-a-ge) [inline exercises/bowling/bowling_test.py]: Could you please also leave a comment stating what version of `canonical-data.json` the tests were adopted as discussed in #784 if aplicable?
The format is:
```text
# Tests adapted from `problem-specifications//canonical-data.json` @ v1.0.0
```

File: exercises/bowling/bowling_test.py [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+class BowlingTests(unittest.TestCase):
+    def test_should_be_able_to_score_a_game_with_all_zeros(self):
+        rolls = [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
+
+        self.game = BowlingGame()
```
Comment (m-a-ge) [inline exercises/bowling/bowling_test.py]: I would also suggest introduce a `setUp` method and create an instance of `BowlingGame` only once
```python
def setUp(self):
        self.game= BowlingGame()
```

File: exercises/bowling/example.py [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+class BowlingGame(object):
+    def __init__(self):
+        self.rolls = []
+        self.totalScore = 0
+        self.currentFrame = Frame()
+        self.bonusRollsAccrued = 0
+        self.bonusRollsSeen = 0
+
+    def roll(self, pins):
+        if (self.isBonusRoll()):
```
Comment (m-a-ge) [inline exercises/bowling/example.py]: probably just:
```python
if self.isBonusRoll():
    # do stuff
```
This increases readability

File: exercises/bowling/example.py [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+            self.bonusRollsSeen += 1
+
+        # is the second roll valid based off the first?
+        if (self.currentFrame.isOpen() and
+           self.currentFrame.getFrame()[0] is not None):
+            if(self.currentFrame.getFrame()[0] + pins > MAX_PINS):
+                raise ValueError
+
+        # open a new frame if the last one has been closed
+        if (self.currentFrame.isOpen() is False):
```
Comment (m-a-ge) [inline exercises/bowling/example.py]:
```python
if not self.currentFrame.isOpen():
```

File: exercises/bowling/example.py [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+            self.currentFrame = Frame()
+
+        # valid roll between 0-10
+        if (pins in range(MAX_PINS + 1)):
+            # raise an error if the game is over and they try to roll again
+            if((len(self.rolls) == NUM_FRAMES) and
+               self.bonusRollsAccrued == 0):
+                raise IndexError
```
Comment (m-a-ge) [inline exercises/bowling/example.py]:
```python
raise IndexError('Failure reason + other details')
```

File: exercises/bowling/example.py [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+            else:
+                self.currentFrame.roll(pins,
+                                       self.isBonusRoll(),
+                                       self.bonusRollsAccrued,
+                                       self.bonusRollsSeen)
+                # if we closed it add it to our rolls
+                if (self.currentFrame.isOpen() is False):
```
Comment (m-a-ge) [inline exercises/bowling/example.py]:
```python
if not self.currentFrame.isOpen():
```

File: exercises/bowling/example.py [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+                    self.rolls.append(self.currentFrame)
+                    # If this is the last frame did we earn any bonus rolls?
+                    if (len(self.rolls) == NUM_FRAMES):
+                        self.bonusRollsEarned()
+                        print(self.bonusRollsAccrued)
+        else:
+            raise ValueError
```
Comment (m-a-ge) [inline exercises/bowling/example.py]:
```python
raise ValueError('Failure details here')
```
The same in other cases

File: exercises/bowling/example.py [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+            frame = self.rolls[frame_index].getFrame()
+
+            roll1 = frame[0]
+            roll2 = frame[1]
+
+            if (self.isStrike(roll1)):
+                self.totalScore += roll1 + self.stikeBonus(frame_index)
+
+            else:
+
```
Comment (m-a-ge) [inline exercises/bowling/example.py]: Unnecessary blank line, please remove

File: exercises/bowling/example.py [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+                        self.bonusRollsEarned()
+                        print(self.bonusRollsAccrued)
+        else:
+            raise ValueError
+
+    def score(self):
+        frame_index = 0
+
+        while (frame_index <= NUM_FRAMES-1):
+            print()
```
Comment (m-a-ge) [inline exercises/bowling/example.py]: Unnecessary statement, please remove

File: exercises/bowling/example.py [hunk 11]
```diff
@@ -0,0 +1,157 @@
...
+                                       self.spareBonus(frame_index)
+                else:
+                    self.totalScore += roll1 + roll2
+
+            frame_index += 1
+
+        return self.totalScore
+
+    def isStrike(self, pins):
+        if (pins == MAX_PINS):
```
Comment (m-a-ge) [inline exercises/bowling/example.py]: It is not necessary to use parenthesis in `if` statements like in `Java`. Add them only to change order precedence or reduce ambiguity

File: exercises/bowling/example.py [hunk 12]
```diff
@@ -0,0 +1,157 @@
...
+            return True
+        else:
+            return False
+
+    def isSpare(self, roll1, roll2):
+        if (roll1 + roll2 == MAX_PINS):
```
Comment (m-a-ge) [inline exercises/bowling/example.py]: Can be simplified to oneliner:
```python
return True if roll1+roll2 == MAX_PINS else False
```
the same in other cases

File: exercises/bowling/example.py [hunk 13]
```diff
@@ -0,0 +1,157 @@
...
+                # may need to close bonus roll frames before 2 have been seen
+                if (bonusRoll and seenBonuses == accruedBonuses):
+                    self.rolls[1] = 0
+                    self.open = False
+            else:
+                # second roll, closes frame
+                self.rolls[1] = roll
+                self.open = False
+
+    def isOpen(self):
```
Comment (m-a-ge) [inline exercises/bowling/example.py]: this doesn't look right, method name doesn't correspond to its actual function

File: exercises/bowling/example.py [hunk 14]
```diff
@@ -0,0 +1,157 @@
...
+    def isBonusRoll(self):
+        # if we've already seen all
+        if (len(self.rolls) >= NUM_FRAMES):
+            return True
+        else:
+            return False
+
+
+class Frame(object):
+
```
Comment (m-a-ge) [inline exercises/bowling/example.py]: It would be great to add docstring

File: exercises/bowling/example.py [hunk 15]
```diff
@@ -0,0 +1,157 @@
...
+                self.bonusRollsAccrued = 2
+            elif (self.isSpare(lastFrame[0], lastFrame[1])):
+                self.bonusRollsAccrued = 1
+        else:
+            self.bonusRollsAccrued = 0
+        return
+
+    def isBonusRoll(self):
+        # if we've already seen all
+        if (len(self.rolls) >= NUM_FRAMES):
```
Comment (m-a-ge) [inline exercises/bowling/example.py]: refactor to oneliner

File: exercises/bowling/example.py [hunk 16]
```diff
@@ -0,0 +1,157 @@
+MAX_PINS = 10
+NUM_FRAMES = 10
+
+
+class BowlingGame(object):
+    def __init__(self):
```
Comment (m-a-ge) [inline exercises/bowling/example.py]: docstring here would be nice 📓

File: exercises/bowling/bowling_test.py [hunk 17]
```diff
@@ -0,0 +1,261 @@
...
+
+        for roll in rolls:
```
Comment (m-a-ge) [inline exercises/bowling/bowling_test.py]: Probably something like:
```python
[self.game,roll(roll) for roll in rolls]
```
And the code is duplicated in every test which is not a good thing, should be refactored

File: exercises/bowling/bowling_test.py [hunk 18]
```diff
@@ -0,0 +1,261 @@
+import unittest
+
+from example import BowlingGame
```
Comment (cmccandless) [inline exercises/bowling/bowling_test.py]: should be `from bowling import BowlingGame`

**Comment by aes421** [comment]
Thanks for the feedback! I just got back from a holiday, and I'm hoping to have time to work on finishing this one later this week.

**Comment by aes421** [comment]
I think I've got everything fixed up.  Let me know what you think. @m-a-ge  @cmccandless

File: exercises/bowling/bowling_test.py [hunk 19]
```diff
@@ -1,258 +1,189 @@
+# Tests adapted from `problem-specifications//canonical-data.json` @ v1.0.1
```
Comment (cmccandless) [inline exercises/bowling/bowling_test.py]: For consistency with the other exercises, could you please move this line like so:
```Python
from bowling import BowlingGame


# Tests adapted from `problem-specifications//canonical-data.json` @ v1.0.1

class BowlingTests(unittest.TestCase):
...
```
Please note the number of blank lines before and after.

File: exercises/bowling/example.py (hunk 13 above)
Comment (aes421) [reply exercises/bowling/example.py]: I'm not sure what you mean on this one.  you call isOpen to find out if the frame is open (true) or not (false)

File: exercises/bowling/bowling_test.py (hunk 19 above)
Comment (aes421) [reply exercises/bowling/bowling_test.py]: done!

**Comment by cmccandless** [comment]
@aes421 Merged; thanks for working on this!

**Pull request merged**